CRM/MembershipExtras/Form/CancelRecurringContribution.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+/**
+ * QuickForm used to cancel memberships.
+ */
+class CRM_MembershipExtras_Form_CancelRecurringContribution extends CRM_Core_Form {
+
+  /**
+   * ID of recurring contribution to be cancelled.
+   *
+   * @var int
+   */
+  private $id;
+
+  /**
+   * ID of contact from where contributions is being cancellsd.
+   *
+   * @var int
+   */
+  private $contactID;
+
+  /**
+   * @inheritdoc
+   */
+  public function preProcess() {
+    $this->id = CRM_Utils_Request::retrieve('crid', 'Positive', $this);
+    $this->contactID = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function buildQuickForm() {
+    CRM_Utils_System::setTitle(ts('Payment Plan Settings'));
+
+    $this->add(
+      'checkbox',
+      'cancel_pending_installments',
+      ts('Do you wish to cancel any pending instalment contribution?'),
+      '',
+      FALSE
+    );
+
+    $this->add(
+      'checkbox',
+      'cancel_memberships',
+      ts('Do you wish to cancel any linked membership?'),
+      '',
+      FALSE
+    );
+
+    $this->addButtons([
+      [
+        'type' => 'submit',
+        'name' => ts('Yes'),
+        'isDefault' => TRUE,
+      ],
+      [
+        'type' => 'cancel',
+        'name' => ts('No'),
+      ],
+    ]);
+  }
+
+  /**
+   * @@inheritdoc
+   */
+  public function postProcess() {
+    $submittedValues = $this->controller->exportValues($this->_name);
+
+    if ($submittedValues['cancel_memberships']) {
+      $this->cancelMemberships();
+    }
+
+    if ($submittedValues['cancel_pending_installments']) {
+      $this->cancelPendingInstallments();
+    }
+
+    $this->cancelRecurringContribution();
+  }
+
+  /**
+   * Cancels memberships being payed for with the current recurring
+   * contribution.
+   */
+  private function cancelMemberships() {
+    civicrm_api3('MembershipPayment', 'get', [
+      'sequential' => 1,
+      'contribution_id.contribution_recur_id.id' => $this->id,
+      'options' => ['limit' => 0],
+      'api.Membership.create' => [
+        'id' => '$value.membership_id',
+        'is_override' => 1,
+        'status_id' => 'Cancelled',
+      ],
+    ]);
+  }
+
+  /**
+   * Cancels pending contributions associated to current recurring contribution.
+   */
+  private function cancelPendingInstallments() {
+    civicrm_api3('Contribution', 'get', [

```

Comment:
I am wondering, is canceling a contribution only change the contribution status (alongside updating the cancle_date and cancle_reason) or is there maybe some other stuff  that civi do in the background ? If you did not do that already, I think it is better to check the queries happen when canceling a membership and ensure that it is similar to what you did below (you can use the tip I shared in email last week to check all queries happen during a certain operation in case u don't know it already).

Reply:
I see, and since you are using cancel API action on ContributionRecur endpoint it will take care of the stuff above. but there is something that we might might need to ask @guanhuan about which is : 

so If I cancel a payment plan (recurring contribution) that has 12 installments attached to it, should the records above get generated only for the recurring contribution or should they get generated also for all the payment plan contributions ?

Reply:
The records above are for payment plan installments exclusively.